src/V3MergeCond.cpp

```diff
@@ -0,0 +1,321 @@
+    // NODE STATE
+    // AstVar::user1 -> Flag set for variables referenced by m_mgCondp
+    AstUser1InUse m_user1InUse;
+
+    // STATE
+    VDouble0 m_statMerges;  // Statistic tracking
+    VDouble0 m_statMergedItems;  // Statistic tracking
+    VDouble0 m_statLongestList;  // Statistic tracking
+
+    AstNode* m_mgFirstp;  // First node in merged sequence
+    AstNode* m_mgCondp;  // The condition of the first node
+    AstNode* m_mgLastp;  // Last node in merged sequence
+    const AstNode* m_mgNextp;  // Next node in list being examined
+    uint32_t m_listLenght;  // Length of current list
+
+    CheckMergeableVisitor m_checkMergeable;  // Sub visitor for encapsulation & speed
+    MarkVarsVisitor m_markVars;  // Sub visitor for encapsulation & speed
+
+    // METHODS
+    VL_DEBUG_FUNC;  // Declare debug()
+
+    // This function extracts the Cond node from the RHS, if there is one and
+    // it is in a supported position, which are:
+    // - RHS is the Cond
+    // - RHS is And(Const, Cond). This And is inserted often by V3Clean.
+    AstNodeCond* extractCond(AstNode* rhsp) {
+        if (AstNodeCond* const condp = VN_CAST(rhsp, NodeCond)) {
+            return condp;
+        } else if (AstAnd* const andp = VN_CAST(rhsp, And)) {
+            if (AstNodeCond* const condp = VN_CAST(andp->rhsp(), NodeCond)) {
+                if (VN_IS(andp->lhsp(), Const)) { return condp; }
+            }
+        }
+        return NULL;
+    }
+
+    // Apply (_ & 1'b1), iff wider than 1 bit
+    AstNode* maskLsb(AstNode* nodep) {
+        if (nodep->width1()) {
+            return nodep;
+        } else {
+            AstNode* const maskp = new AstConst(nodep->fileline(), AstConst::LogicTrue());
+            return new AstAnd(nodep->fileline(), nodep, maskp);
+        }
+    }
+
+    // Fold the RHS expression assuming the given condition state. Unlink bits
+    // from the RHS which is only used once, and can be reused. What remains
+    // of the RHS is expected to be deleted by the caller.
+    AstNode* foldAndUnlink(AstNode* rhsp, bool condTrue) {
+        if (rhsp->sameTree(m_mgCondp)) {
+            return condTrue ? new AstConst(rhsp->fileline(), AstConst::LogicTrue())
+                            : new AstConst(rhsp->fileline(), AstConst::LogicFalse());
+        } else if (AstNodeCond* const condp = extractCond(rhsp)) {
+            AstNode* const resp
+                = condTrue ? condp->expr1p()->unlinkFrBack() : condp->expr2p()->unlinkFrBack();
+            if (condp == rhsp) { return resp; }
+            if (AstAnd* const andp = VN_CAST(rhsp, And)) {
+                UASSERT_OBJ(andp->rhsp() == condp, rhsp, "Should not try to fold this");
+                return new AstAnd(andp->fileline(), andp->lhsp()->cloneTree(false), resp);
+            }
+        } else if (AstAnd* const andp = VN_CAST(rhsp, And)) {
+            if (andp->lhsp()->sameTree(m_mgCondp)) {
+                return condTrue ? maskLsb(andp->rhsp()->unlinkFrBack())
```

Comment:
When is maskLsb needed?  I would think you either 1. you don't need the And mask, 2. or if you did you'd have to look at every bit.

Reply:
I didn't think I need that either until EH2 broke... Sometimes you have an And(1-bit-value, more-than-1-bit-value) on the RHS, which I think is also just strength reduced form of `cond ? x[0] : 1'b0` and similar. This test covers it and fails without:
```
      // Merge with an '&' also used for masking of LSB.
      shuf_e[0] = cond3 ? bits[0] : bits[63];
      for (int i = 1; i < 64; i = i + 1) begin
         shuf_e[i] = cond3 & crc[0];
      end
```
Without the And inserted by MergeCond  you would assign `shuf_e = crc` on the true path, but the rest of the code assumes shuf_e is a clean 1-bit thing so junk in the top causes boom.

Reply:
Ah, fallout of this optimization being after V3Clean which optimizes the upper bits and normally does this sort of And-ing.

Reply:
Exactly

Reply:
P.S. Probably worth a comment there about this case.